boot/freeldr/freeldr/arch/i386/multiboot.S

```diff
@@ -139,32 +129,38 @@ mb1:
     jnz mb1
 
 mb2:
+#ifndef _M_IX86
+/* If the compiler is 64 bits then skip the 32 bits prefix using near jump we are already in protected mode */
+    .byte HEX(E9)
+    .long 1
+#endif
     /* See if the boot device was passed in */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_BOOT_DEVICE
+    test dword ptr [ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_BOOT_DEVICE
 
     /* If no boot device known, assume first partition of first harddisk */
     mov dx, HEX(0180)
     jz mb3
 
+#ifndef _M_IX86
+/* If the compiler is 64 bits then skip the 32 bits prefix using near jump we are already in protected mode */
+    .byte HEX(E9)
+    .long 1
+#endif
     /* Load boot drive into DL, boot partition into DH */
-    mov edx, dword ptr ds:[ebx + MB_INFO_BOOT_DEVICE_OFFSET]
+    mov edx, dword ptr [ebx + MB_INFO_BOOT_DEVICE_OFFSET]
     bswap edx
-    inc dh
 
 mb3:
     /* Relocate itself to lower address */
     mov esi, INITIAL_BASE
     mov edi, FREELDR_BASE
-    mov ecx, offset __bss_start__ - FREELDR_BASE
-    shr ecx, 2
-    rep movsd
+    mov ecx, offset __bss_end__ - FREELDR_BASE
+    rep movsb
 
     /* Load segment registers for real-address mode */
-#ifdef _USE_ML
-    lgdt fword ptr ds:[gdtptr]
-#else
-    lgdt ds:[gdtptr]
-#endif
+    lgdt32A gdtptr
+    lidt32A idtptr
```

Comment:
```suggestion
    lgdt [gdtptr]
    lidt [idtptr]
```

Reply:
i made something compatible with both GAS and ML.